R-package/src/cmake/modules/FindLibR.cmake

```
@@ -0,0 +1,199 @@
+# CMake module used to find the location of R's
+# dll and header files.
+#
+# Borrows heavily from xgboost's R package:
+#
+# * https://github.com/dmlc/xgboost/blob/master/cmake/modules/FindLibR.cmake
+#
+# Defines the following:
+#  LIBR_FOUND
+#  LIBR_HOME
+#  LIBR_EXECUTABLE
+#  LIBR_INCLUDE_DIRS
+#  LIBR_LIB_DIR
+#  LIBR_CORE_LIBRARY
+# and a CMake function to create R.lib for MSVC
+
+if(NOT R_ARCH)
+  if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "4")
+    set(R_ARCH "i386")
+  else()
+    set(R_ARCH "x64")
+  endif()
+endif()
+
+if(NOT ("${R_ARCH}" STREQUAL "x64"))
+  message(FATAL_ERROR "LightGBM's R package currently only supports 64-bit operating systems")
+endif()
+
+# Creates R.lib and R.def in the build directory for linking with MSVC
+function(create_rlib_for_msvc)
+
+  message("Creating R.lib and R.def")
+
+  # various checks and warnings
+  if(NOT WIN32 OR NOT MSVC)
+    message(FATAL_ERROR "create_rlib_for_msvc() can only be used with MSVC")
+  endif()
+
+  if(NOT EXISTS "${LIBR_LIB_DIR}")
+    message(FATAL_ERROR "LIBR_LIB_DIR, '${LIBR_LIB_DIR}', not found")
+  endif()
+
+  find_program(GENDEF_EXE gendef)
+  find_program(DLLTOOL_EXE dlltool)
+
+  if(NOT GENDEF_EXE OR NOT DLLTOOL_EXE)
+    message(FATAL_ERROR "Either gendef.exe or dlltool.exe not found!\
+      \nDo you have Rtools installed with its MinGW's bin/ in PATH?")
+  endif()
+
+  # extract symbols from R.dll into R.def and R.lib import library
+  execute_process(COMMAND ${GENDEF_EXE}
+    "-" "${LIBR_LIB_DIR}/R.dll"
+    OUTPUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+  )
+  execute_process(COMMAND ${DLLTOOL_EXE}
+    "--input-def" "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib"
+  )
+endfunction(create_rlib_for_msvc)
+
+# R version information is used to search for R's libraries in
+# the registry on Windows. Since this code is orchestrated by
+# an R script (src/install.libs.R), that script uses R's built-ins to
+# find the version of R and pass it through as a CMake variable
+if(CMAKE_R_VERSION)
+  message("R version passed into FindLibR.cmake: ${CMAKE_R_VERSION}")
+else()
+  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in but none was provided. Check src/install.libs.R")
```

I think we can require ``R_VERSION`` only on Windows for the sake of general usage of FindLibrR.
```suggestion
elseif(WIN32)
  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in on Windows but none was provided. Check src/install.libs.R")
```